File: PredictingGeneExpression.R
library(caret)
library(GGally)
library(ggplot2)
library(txtplot)

source("modelpipe.R")
source("MaclearnUtilities.R")


## -----------------------------------------------------------------
## load Patel data
## -----------------------------------------------------------------
readTab = function(file) {
	if (grepl("gz$", file)) {
		file = gzfile(file)
	}
	read.table(file, sep="\t",
			header=TRUE, row.names=1, check.names=FALSE)
}

x = data.frame(t(readTab(
        "rnaseq/GSE57872/GSE57872_DataMatrixMapped.tsv.gz")),
        check.names=FALSE)
y = x$BRCA1
names(y) = rownames(x)
x0 = x[ , colnames(x) != "BRCA1"]

corPVals = apply(x0, 2,
        function(z) {cor.test(z, y)$p.value})
corQVals = p.adjust(corPVals, method="fdr")
head(sort(corQVals))
summary(lm(y ~ x0$CDK1))

ggobj = ggplot(data=x, mapping=aes(x=CDK1, y=BRCA1))
ggobj = ggobj + theme_classic()
ggobj = ggobj + geom_point(alpha=0.5)
ggobj = ggobj + stat_smooth(method="lm", se=FALSE)
print(ggobj)


## -----------------------------------------------------------------
## unregularized linear regression
## -----------------------------------------------------------------
nFeats = c(2, 5, 10, 20, 50, 100, 200, 500, 1000)
names(nFeats) = as.character(nFeats)
brca1Modelers = lapply(X=nFeats, FUN=function(n) {
    solder(
        PearsonSelector(nFeat = n),
        GlmFitter(fam="gaussian", alpha=0, lambda=0)
    )
})

brca1Model20 = brca1Modelers[["20"]](x0, y)
brca1Preds = predict(brca1Model20, x0)
txtplot(brca1Preds, y)

brca1Model1000 = brca1Modelers[["1000"]](x0, y)
brca1Preds = predict(brca1Model1000, x0)
txtplot(brca1Preds, y)

set.seed(123)
brca1CV = lapply(X=brca1Modelers, FUN=function(m) {train(m, x0, y)})

ggdata = data.frame(
    "Number Potential Features" = nFeats,
    Rsquared = sapply(brca1CV, function(m) {m$results$Rsquared}),
    Regularization = "none",
    Lambda = 0,
    check.names = FALSE
)
ggobj = ggplot(
    data = ggdata,
    mapping = aes(x=`Number Potential Features`, y=Rsquared)
)
ggobj = ggobj + theme_classic()
ggobj = ggobj + geom_point() + geom_line(alpha=0.5)
ggobj = ggobj + scale_x_log10()
print(ggobj)



## -----------------------------------------------------------------
## L2-regularized linear regression
## -----------------------------------------------------------------
brca1Modelers2 = lapply(X=nFeats, FUN=function(n) {
    solder(
        PearsonSelector(nFeat = n),
        GlmFitter(fam="gaussian", alpha=0, lambda=NULL)
    )
})

set.seed(123)
brca1CV2 = lapply(X=brca1Modelers2,
		FUN=function(m) {train(m, x0, y)})

lambdaMins2 = sapply(brca1CV2, FUN=function(z) {
    z$finalModel[[2]]$fit$lambda.min
})
summary(lm(lambdaMins2 ~ nFeats))
## approx 1.5 + 0.034 * nFeats

ggdata = rbind(ggdata, data.frame(
    "Number Potential Features" = nFeats,
    Rsquared = sapply(brca1CV2, function(m) {m$results$Rsquared}),
    Regularization = "L2/ridge",
    Lambda = lambdaMins2,
    check.names = FALSE
))


## -----------------------------------------------------------------
## L1-regularized linear regression
## -----------------------------------------------------------------
brca1Modelers1 = lapply(X=nFeats, FUN=function(n) {
    solder(
        PearsonSelector(nFeat = n),
        GlmFitter(fam="gaussian", alpha=1, lambda=NULL)
    )
})

set.seed(123)
brca1CV1 = lapply(X=brca1Modelers1,
		FUN=function(m) {train(m, x0, y)})

lambdaMins1 = sapply(brca1CV1, FUN=function(z) {
    z$finalModel[[2]]$fit$lambda.min
})
summary(lm(lambdaMins1 ~ log(nFeats)))
## approx -0.0157 + 0.0235 * log(nFeats)

ggdata = rbind(ggdata, data.frame(
    "Number Potential Features" = nFeats,
    Rsquared = sapply(brca1CV1, function(m) {m$results$Rsquared}),
    Regularization = "L1/Lasso",
    Lambda = lambdaMins1,
    check.names = FALSE
))

ggobj = ggplot(
    data = ggdata,
    mapping = aes(x=`Number Potential Features`, y=Rsquared,
            linetype=Regularization)
)
ggobj = ggobj + theme_classic()
ggobj = ggobj + geom_point() + geom_line(alpha=0.5)
ggobj = ggobj + scale_x_log10()
print(ggobj)

File: PredictingGeneExpression.py
from collections import OrderedDict
import copy
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from pandas import DataFrame, Series
import plotnine as gg
import scipy as sp
import scipy.stats as stats
import sklearn as sk
import sklearn.model_selection as model_selection
from sklearn.model_selection import ShuffleSplit
import sklearn.feature_selection as feature_selection
import sklearn.linear_model as linear_model
import sklearn.pipeline as pipeline

import MaclearnUtilities
from MaclearnUtilities import bhfdr, colcor

plt.ion()

def pandaize(f):
    def pandaized(estimator, X, y, **kwargs):
        return f(estimator, np.array(X), y, **kwargs)
    return pandaized

@pandaize
def cross_val_score_pd(estimator, X, y, **kwargs):
    return model_selection.cross_val_score(estimator, X, y, **kwargs)


## -----------------------------------------------------------------
## load Patel data
## -----------------------------------------------------------------
def readTab(file):
    return pd.read_csv(file, sep="\t", header=0, index_col=0)

x = readTab("rnaseq/GSE57872/GSE57872_DataMatrixMapped.tsv.gz").transpose()
y = x.BRCA1
x0 = x[ x.columns[x.columns != "BRCA1"] ]

cvSched = ShuffleSplit(n_splits=10, test_size=0.1, random_state=123)

corPVals = colcor(x0, y)['p']
corQVals = bhfdr(corPVals)
corQVals.sort_values(inplace=False).head()

plt.close()
ax = plt.subplot(111)
x.plot.scatter(x="CDK1", y="BRCA1", ax=ax)


## -----------------------------------------------------------------
## unregularized linear regression
## -----------------------------------------------------------------
nFeats = [2, 5, 10, 20, 50, 100, 200, 500, 1000]
brca1Modelers = OrderedDict([
    (n, pipeline.Pipeline([
        ('featsel', feature_selection.SelectKBest(
                feature_selection.f_regression, k=n)),
        ('regressor', linear_model.LinearRegression())
    ]))
    for n in nFeats
])
            
brca1Model20 = copy.deepcopy(brca1Modelers[20]).fit(x0, y)
brca1Preds = brca1Model20.predict(x0)
stats.pearsonr(brca1Preds, y)[0]

brca1Model1000 = copy.deepcopy(brca1Modelers[1000]).fit(x0, y)
brca1Preds = brca1Model1000.predict(x0)
stats.pearsonr(brca1Preds, y)[0]

cvR2s_unreg = Series(OrderedDict([
    (n, np.mean(cross_val_score_pd(copy.deepcopy(brca1Modelers[n]),
                                   X = x0,
                                   y = y,
                                   cv = cvSched.split(x0))))
    for n in nFeats
]))


## -----------------------------------------------------------------
## L2-regularized linear regression
## -----------------------------------------------------------------
brca1Modelers2 = OrderedDict([
    (n, pipeline.Pipeline([
        ('featsel', feature_selection.SelectKBest(
                feature_selection.f_regression, k=n)),
        ('regressor', linear_model.Ridge(
                alpha=len(y)*(1.5 + 0.034*n)))
    ]))
    for n in nFeats
])
            
cvR2s_L2 = Series(OrderedDict([
    (n, np.mean(cross_val_score_pd(copy.deepcopy(brca1Modelers2[n]),
                                   X = x0,
                                   y = y,
                                   cv = cvSched)))
    for n in nFeats
]))


## -----------------------------------------------------------------
## L1-regularized linear regression
## -----------------------------------------------------------------
brca1Modelers1 = OrderedDict([
    (n, pipeline.Pipeline([
        ('featsel', feature_selection.SelectKBest(
                feature_selection.f_regression, k=n)),
        ('regressor', linear_model.Lasso(
                alpha=max(0, (0.0235*np.log(n)-0.0157))))
    ]))
    for n in nFeats
])
            
cvR2s_L1 = Series(OrderedDict([
    (n, np.mean(cross_val_score_pd(copy.deepcopy(brca1Modelers1[n]),
                                   X = x0,
                                   y = y,
                                   cv = cvSched)))
    for n in nFeats
]))


## -----------------------------------------------------------------
## plot results
## -----------------------------------------------------------------
plotdata = DataFrame({
    "Number Potential Features" : nFeats * 3,
    "Rsquared" : pd.concat([cvR2s_unreg, cvR2s_L2, cvR2s_L1]),
    "Regularization" : (['-']*len(nFeats) +
                        ['L2/ridge']*len(nFeats) +
                        ['L1/lasso']*len(nFeats))
})
plotdata.index = (plotdata["Number Potential Features"].apply(str) + "_" +
                plotdata["Regularization"])
plotdata = plotdata.loc[plotdata.Rsquared > 0]

plt.close()
ggo = gg.ggplot(plotdata, gg.aes(x = 'Number Potential Features',
                                 y = 'Rsquared',
                                 linetype = 'Regularization'))
ggo += gg.geom_line()
ggo += gg.theme_bw()
ggo += gg.scale_x_log10()
print(ggo)

